server/src/main/java/org/elasticsearch/search/dfs/DfsPhase.java

```diff
@@ -171,34 +172,36 @@ public Query rewrite(Query original) throws IOException {
 
     private void executeKnnVectorQuery(SearchContext context) throws IOException {
         SearchSourceBuilder source = context.request().source();
-        if (source == null || source.knnSearch() == null) {
+        if (source == null || source.knnSearch() == null || source.knnSearch().isEmpty()) {
             return;
         }
 
         SearchExecutionContext searchExecutionContext = context.getSearchExecutionContext();
-        KnnSearchBuilder knnSearch = context.request().source().knnSearch();
-        KnnVectorQueryBuilder knnVectorQueryBuilder = knnSearch.toQueryBuilder();
+        List<KnnSearchBuilder> knnSearch = context.request().source().knnSearch();
+        List<KnnVectorQueryBuilder> knnVectorQueryBuilders = knnSearch.stream().map(KnnSearchBuilder::toQueryBuilder).toList();
 
         if (context.request().getAliasFilter().getQueryBuilder() != null) {
-            knnVectorQueryBuilder.addFilterQuery(context.request().getAliasFilter().getQueryBuilder());
+            for (KnnVectorQueryBuilder knnVectorQueryBuilder : knnVectorQueryBuilders) {
+                knnVectorQueryBuilder.addFilterQuery(context.request().getAliasFilter().getQueryBuilder());
+            }
         }
-
-        Query query = searchExecutionContext.toQuery(knnVectorQueryBuilder).query();
-        TopScoreDocCollector topScoreDocCollector = TopScoreDocCollector.create(knnSearch.k(), Integer.MAX_VALUE);
-        Collector collector = topScoreDocCollector;
-
-        if (context.getProfilers() != null) {
-            InternalProfileCollector ipc = new InternalProfileCollector(
-                topScoreDocCollector,
-                CollectorResult.REASON_SEARCH_TOP_HITS,
-                List.of()
-            );
-            context.getProfilers().getDfsProfiler().setCollector(ipc);
-            collector = ipc;
+        List<DfsKnnResults> knnResults = new ArrayList<>(knnVectorQueryBuilders.size());
+        for (int i = 0; i < knnSearch.size(); i++) {
+            Query knnQuery = searchExecutionContext.toQuery(knnVectorQueryBuilders.get(i)).query();
+            TopScoreDocCollector topScoreDocCollector = TopScoreDocCollector.create(knnSearch.get(i).k(), Integer.MAX_VALUE);
+            Collector collector = topScoreDocCollector;
+            if (context.getProfilers() != null) {
+                InternalProfileCollector ipc = new InternalProfileCollector(
+                    topScoreDocCollector,
+                    CollectorResult.REASON_SEARCH_TOP_HITS,
+                    List.of()
+                );
+                context.getProfilers().getDfsProfiler().setCollector(ipc);
+                collector = ipc;
+            }
+            context.searcher().search(knnQuery, collector);
+            knnResults.add(new DfsKnnResults(topScoreDocCollector.topDocs().scoreDocs));
```

Comment:
I was hoping here to be able to do some `boolean` query with a `should`, but I don't think there is a way to determine which clauses contributed to a document's score. If we could that, we can get away with just one query here instead the number of queries per knn clause.

Reply:
Where is the distinction needed? I was under the impression that we merge all results into the same score doc array anyways. I guess the challenge is to sum the two scores, or is there more to it?
Is it necessary to carry around a list of dfs knn results? Would it be an option to merge them into a single array here, and possibly remove duplicates here too rather than later?

Reply:
> Where is the distinction needed? 

This is because each clause could have a different `K`. So, we need to separate them out when shipped back to the coordinator.

Reply:
Right. The approach is sensible. We may want to look at running these in parallel instead of sequentially.